### ISSUE TITLE
Create a pipeline for PAC in Jenkins using JobDSL

### ISSUE THREAD

**Issue by MadsNielsen** [Praqma/Praqmatic-Automated-Changelog#17]
We want to have a pipeline setup in Jenkins - The pipeline should be a simple one. We want it to be flexible enough so that i can be re-used for different branches. 

The pipeline, and sequence of jobs. should be created using JobDSL.

We already have the rake tasks in place that run the tests in our suite. So let's include these rake tasks in our pipeline in our test job.

```
task :default => [:test]

task :test do
	Dir.glob('test/unit/*.rb').each do |testfile|
		ruby "#{testfile}"
	end
end

task :functional_test do 
	Dir.glob('test/functional/*.rb').each do |testfile|
		ruby "#{testfile}"
	end
end

task :coverage do
  ENV['COVERAGE'] = 'on'
  Rake::Task['test'].execute
end
```

Examples of how DSL can look like can be found here. The pipeline scripts with helper scritps should be OK.

https://github.com/Praqma/job-dsl-collection

**Comment by buep** [comment]
Adding to milestone - part of our standard delivery.

**Comment by Ordojan** [comment]
@buep Can you guys check my script out, first time I'm creating something with jobdsl. I planned to add more features, but had to compromise when I realised that I won't have enough time.

**Comment by buep** [comment]
Feedback:

* github url is wrong
* the release job shouldn't take parameters, I suggest you source a version.sh file instead and read eg. VERSION=2.0.0 and then add buildnumber so version becomes 2.0.0-BN
* first job have the a pretested integration job, we follow this workflow in all our development: http://www.praqma.com/stories/a-pragmatic-workflow/ so you will configure first job to be a pretested integration job running build (there acutally aren't a build..) but then test
* the next job is integration tests
* third job is release, and that do not take any parameters from manual input

You also need to place the jobs on our Jenkins, and make the seed job.
You should ask @praqma-thi  how they do it for the plugin development, and follow same principles.

**Comment by buep** [comment]
And ask them also where the actually place the DSL files... what are our conventions for that? Is that really to put them in the directory: jenkins-pipeline/pipeline.groovy ?

**Comment by Ordojan** [comment]
@buep @MadsNielsen  I think it's done. 

[here ](https://github.com/Praqma/Praqmatic-Automated-Changelog/blob/2.0/jenkins-pipeline/pipeline.groovy) is the script. 

Seed [job](http://code.praqma.net/ci/view/seed/job/PAC-pipeline-seed/).

[Here ](http://code.praqma.net/ci/view/Open%20Source/view/Praqmatic-Automated-Changelog/)is the pipeline.

Pipeline [view](http://code.praqma.net/ci/view/Open%20Source%20Pipelines/view/Praqmatic-Automated-Changelog%20pipeline/).

Check it out and tell me if it's correct.

**Comment by buep** [comment]
Feedback:

* your jobs are not guaranteed to run on the same git sha. It must be passed downstream using trigger paramerize downstream job from second job and forward.
* you also hardcoded many values several times... I like them to be placed in the top of the groovy script.

Further there are some nice to have stuff like showing sha build, setting build description etc.

Assign to me a let me look into it. I can easily fix it.

**Comment by buep** [comment]
@Ordojan - did you try to run the jobs?

**Comment by Ordojan** [comment]
ummm yeah...

**Comment by buep** [comment]
Here are the things that needs to be fixed, together with few tips and tricks:

* configure project-based permissions, so public can read the pipeline jobs. E.g. like this job: http://code.praqma.net/ci/view/Open%20Source%20Pipelines/view/Pretested%20Integration%20Plugin%20-%20build%20pipeline/job/pretested-integration-plugin_push-Github-JenkinsCI_master/configure
* any value in the job dsl that is mentioned more than once should be defined in the top so we only have to change it once
* use execute shell build steps, and not specific rake tasks that depends on the rake plugin. Use shell build steps to make sure it works the same way as on command line not being wrapped in a plugin
* your jobs are not guaranteed to run on the same git sha. It must be passed downstream using trigger paramerize downstream job from second job and forward. - see attached picture for an example
![selection_053](https://cloud.githubusercontent.com/assets/1579826/14428175/40bbfe20-fff8-11e5-89b3-443cb44b9ee3.png)
* Also set the buildname to be: #${BUILD_NUMBER}#${GIT_REVISION,length=8}(${GIT_BRANCH})
* make sure the job always run on appropriate slave
* remove me from mail notifications
* make sure the release job also mentions the version number in the buildname
* ensure that when a release build is executed, the complete pipeline (all the jobs in the line) is marked Keep Forever, using the promotion plugin.
* increate number of old builds to 100.
* also set the job ownership to be you and Mads (example here: http://code.praqma.net/ci/view/seed/job/moeller-ucd-seed/configure) 
* the test jobs must show test results, if such are produced. It there isn't produced test results, file a issues requesting build process to output them so we can do it later.

Many of the setups are shown and used in this pipeline also: http://code.praqma.net/ci/view/Open%20Source%20Pipelines/view/Pretested%20Integration%20Plugin%20-%20build%20pipeline/

And here are examples on many of the above improvements also: http://code.praqma.net/ci/view/seed/job/cool-pipeline-seed/configure

**Comment by praqma-thi** [comment]
> ensure that when a release build is executed, the complete pipeline (all the jobs in the line) is marked Keep Forever, using the promotion plugin.
The promotion plugin doesn't save much data in the job's config.xml, most of it is stored elsewhere. 

Job DSL support has been merged into their master but it hasn't been released yet. I'd suggest we wait until they release it, because configuring promotions for jobs outscopes our scripts. 

It might be possible to configure it in our scripts since they run on the master? I'd have to look into it. I've some experience with this kind of stuff, but mostly negative experiences, hehe..

**Comment by buep** [comment]
Further job DSL reference for promotion:

* https://github.com/codecentric/job-dsl-promotions-plugin
* https://issues.jenkins-ci.org/browse/JENKINS-21750
* https://github.com/denschu/job-dsl-examples

Okay, wait with the promotion - after an release go in and mark the build keep forever manually.

File an issue that it is missing in the pipeline.

**Comment by buep** [comment]
> the test jobs must show test results, if such are produced. It there isn't produced test results, file a issues requesting build process to output them so we can do it later.

Filed as issue #31

**Comment by buep** [comment]
@Ordojan  - questions:

- Why are you sourcing the user profile in the build step?
- Why is the first job doing nothing? I mean, it would be natural to build or run unit test as criteria for integration. At least run pac.rb -h to see the basic Ruby script is running. Any reason not to?

**Comment by buep** [comment]
This set build won't work:

${BUILD_NUMBER}#${GIT_REVISION,length=8}(${GIT_BRANCH})-ver=$VERSION

See the help text of the plugin, that environment variables like the $VERSION need to be used liked this:

${ENV,var="VARIABLENAME"}
    Expands to an environment variable (specified here as VARIABLENAME) from the build environment. Note that this does not include any variables set by the build scripts themselves, only those set by Jenkins and other plugins.

**Comment by buep** [comment]
@Ordojan - you didn't check the pipeline? The pipeline view wasn't working so pointed to the first job again.

The integration job do not have any downstream job - why is that. If we want to visualize it using the pipeline, I suggest we still up/downstream them but the second job just run master and is triggerede by upstream (integration) job and not polling/push notification.

You didn't fix the env-var name I mentioned above either. Release job will still not show version number or it will fail.

Further you didn't explain why you still source the profile in job nr. 2 - I think I now the explanation, but please add it as a comment in build step.

**Comment by MadsNielsen** [comment]
@buep That with the integration job was actually me. Since the first job (the toll gate) uses pretested integration deletes a branch upon succes...i didn't think you could pass down the sha that the SCM checked out. So that was something i removed?

**Comment by buep** [comment]
You can't pass down the SHA (easily), but we can still trigger it without the SHA and just let it take latest from master? Won't that be okay? Any pitfalls I miss?
Then we have a visual pipeline with all steps, not the first one missing.

**Comment by MadsNielsen** [comment]
@buep Your choice, i prefer it the other way since there will no mistakes and freak accidental timing issues or users doing funny things with git and shell build steps in the interim. Besides, the link from the toll gate (possibly the squashed commits) and the integration is very temporary. We only really care about the commits on master.

**Comment by buep** [comment]
Which mistakes and freak accidents do I miss? Examples please.

I don't see any problems as the job do not run in parallel and they execute in order.

**Comment by buep** [comment]
Is everything done for this issue?

I mean - is the pipeline as we like it to be? Is all things I pointed out still there, and implemented?

Please review the pipeline and check against this issues - then close.

**Comment by Ordojan** [comment]
All the points are covered in the current pipeline. One issue is that at this point it is not possible to use the pipeline view. If that is okay then we can close this issue. Otherwise we need to change a few settings and connect the jobs using BuildTriggers.

**Comment by buep** [comment]
> All the points are covered in the current pipeline. One issue is that at this point it is not possible to use the pipeline view. If that is okay then we can close this issue. Otherwise we need to change a few settings and connect the jobs using BuildTriggers.

Of course this is not okay.

Please go over all my comments, one by one and check that the pipeline and jobs are okay. Do it very thoroughly - don't miss things.
For example the VERSION number in the build name for release job do still not work! I added several comments on that.

Note I didn't run the seed job, I assume the pipeline is up-to-date?

Why is there 4 jobs generated by the seed job? Doesn't it clean up?

Further make sure that the jobs dsl is documented - if you read this issues, there are lots of discussions and reasoning behind our job dsl that should go there as comments inside the job dsl.
Please finish that up as well.

**Comment by Ordojan** [comment]
@buep  I copied the release job and ran it on a fork of the repositoy [here](http://code.praqma.net/ci/view/Open%20Source/view/Praqmatic-Automated-Changelog/job/3_testrelease_pac/) and the releases are [here](https://github.com/Ordojan/Praqmatic-Automated-Changelog/releases). It seems like the version number works (on linux), unless I'm misunderstanding something.

**Comment by buep** [comment]
@praqma-thi - can you go through this issue and check if Andrius is done? I mean all the things discussed and pointed out are fixed as well as all the requirements I mentioned?

I'm tired of re-opening this all the time and want it closed once and for all now.

**Comment by buep** [comment]
And note that Andrius didn't update the issues yesterday, but this comment was cleared out - it works fine:
> @buep I copied the release job and ran it on a fork of the repositoy here and the releases are here. It seems like the version number works (on linux), unless I'm misunderstanding something.

**Comment by praqma-thi** [comment]
The only things missing were done so deliberately:
 * **Build promotion** - Waiting on the plugin's new release. They're actively working on it.
 * **Display test results** - However, results are written to console.  Relates to #31.

Everything else seems to be accounted for.

I made some minor changes to the DSL, fixing some innocent mistakes.
There's still room for improvement, but there always is.

**Comment by buep** [comment]
Okay, I would have preferred a few more explanations in the job dsl as comments, but let's close this now you also agree.

Build promotion is mentioned now in a new issues I wrote #41.

Thanks, you can close it.

**Comment by buep** [comment]
Sorry it is closed  - fine :-)
